db/stats.py

```diff
@@ -1,119 +1,244 @@
+        mbid (MusicBrainz ID), artist (name), and title.
+    """
+    cache_key = "last-submitted-recordings"
+    last_submissions = db.cache.get(cache_key)
+    if not last_submissions:
         with db.engine.connect() as connection:
-            result = connection.execute("""SELECT mbid,
-                                     data->'metadata'->'tags'->'artist'->>0,
-                                     data->'metadata'->'tags'->'title'->>0
-                                FROM lowlevel
-                            ORDER BY id DESC
-                               LIMIT 5
-                              OFFSET 10""")
-            last_submitted_data = result.fetchall()
-        last_submitted_data = [
-            (r[0], r[1], r[2]) for r in last_submitted_data if r[1] and r[2]
+            # We are getting results with of offset of 10 rows because we'd
+            # prefer to show recordings for which we already calculated
+            # high-level data. This might not be the best way to do that.
+            result = connection.execute("""
+                SELECT mbid,
+                       data->'metadata'->'tags'->'artist'->>0,
+                       data->'metadata'->'tags'->'title'->>0
+                  FROM lowlevel
+              ORDER BY id DESC
+                 LIMIT 5
+                OFFSET 10
+            """)
+            last_submissions = result.fetchall()
+        last_submissions = [
+            {
+                "mbid": r[0],
+                "artist": r[1],
+                "title": r[2],
+            } for r in last_submissions if r[1] and r[2]
         ]
-        db.cache.set('last-submitted-data', last_submitted_data, time=LAST_MBIDS_CACHE_TIMEOUT)
+        db.cache.set(cache_key, last_submissions, time=LAST_MBIDS_CACHE_TIMEOUT)
 
-    return last_submitted_data
+    return last_submissions
 
+def get_earliest_submission_date():
+    """Get the earliest date that something was submitted to AB."""
+    q = text("""SELECT submitted
+                  FROM lowlevel
+              ORDER BY submitted ASC
+                 LIMIT 1""")
+    with db.engine.connect() as connection:
+        cur = connection.execute(q)
+        row = cur.fetchone()
+        if row:
+            return row[0]
+
+def get_most_recent_stats_date():
+    q = text("""SELECT collected
+                  FROM statistics
+              ORDER BY collected DESC
+                 LIMIT 1""")
+    with db.engine.connect() as connection:
+        cur = connection.execute(q)
+        row = cur.fetchone()
+        if row:
+            return row[0]
+
+def get_next_hour(date):
+    """Round up a date to the nearest hour:00:00.
+    Arguments:
+      date: a datetime
+    """
+    delta = datetime.timedelta(hours=1)
+    date = date + delta
+    date = date.replace(minute=0, second=0, microsecond=0)
+    return date
+
+def compute_stats(to_date):
+    # Compute stats up to the given date
```

Comment:
yeah, I just pushed this so that I would have it on my other machine. Still some stuff to do